runners/direct-java/src/main/java/org/apache/beam/runners/direct/ParDoEvaluator.java

```diff
@@ -100,7 +101,8 @@
               windowingStrategy,
               doFnSchemaInformation,
               sideInputMapping);
-      if (DoFnSignatures.signatureForDoFn(fn).usesState()) {
+      DoFnSignature signature = DoFnSignatures.signatureForDoFn(fn);
+      if (signature.usesState() || signature.onWindowExpiration() != null) {
```

Comment:
Is this relevant to this change?

Reply:
Yes, though it could also be separated. OnWindowExpiration didn't work if there were no state variables, which caused some of the tests I added to fail.